servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/PublisherBuffer.java

```diff
@@ -249,57 +290,111 @@ public void onComplete() {
-                        @SuppressWarnings("unchecked")
-                        ItemsTerminated<T, B> itemsTerminated = (ItemsTerminated<T, B>) cMaybeAccumulator;
-                        terminateTarget(itemsTerminated.accumulator, target, itemsTerminated.terminalNotification,
-                                bSubscription);
+                    if (maybeAccumulatorUpdater.compareAndSet(this, ADDING,
+                            new NextAccumulatorHolder<>(nextAccumulator))) {
+                        // Thread that is ADDING will observe the change and emit the current accumulator
                         return;
                     }
+                } else if (NextAccumulatorHolder.class.equals(cMaybeAccumulator.getClass())) {
+                    // Hand-off emission of one more nextAccumulator to the thread that is still ADDING, discarding it.
+                    // Since we did not emit the accumulator we request one more to observe the next boundary:
+                    bSubscription.request(1);
+                    return;
                 } else {
-                    assert cMaybeAccumulator instanceof Accumulator;
-                    if (maybeAccumulatorUpdater.compareAndSet(this, cMaybeAccumulator, nextAccumulator)) {
+                    assert cMaybeAccumulator instanceof CountingAccumulator;
+                    final NextAccumulatorHolder<T, B> holder = new NextAccumulatorHolder<>(nextAccumulator);
+                    if (maybeAccumulatorUpdater.compareAndSet(this, cMaybeAccumulator, holder)) {
                         @SuppressWarnings("unchecked")
                         Accumulator<T, B> oldAccumulator = (Accumulator<T, B>) cMaybeAccumulator;
-                        target.onNext(oldAccumulator.finish());
+                        if (deliverOnNext(oldAccumulator, target, bSubscription, tSubscription)) {
+                            final Object nextState = maybeAccumulatorUpdater.accumulateAndGet(this,
+                                    // Keep "prev" state because we observed `itemsTerminated` or `accumulate` took
+                                    // ownership of the `holder`
+                                    toCounting(nextAccumulator), (prev, next) -> prev == holder ? next : prev);
+                            if (ItemsTerminated.class.equals(nextState.getClass())) {
+                                maybeAccumulator = TERMINATED;
+                                @SuppressWarnings("unchecked")
+                                final ItemsTerminated<T, B> it = (ItemsTerminated<T, B>) nextState;
+                                // Deliver the last boundary only if there are some items pending and demand is present
+                                final Accumulator<T, B> accumulator = it.accumulator.size() > 0 &&
+                                        pendingUpdater.decrementAndGet(this) >= 0L ? it.accumulator : null;
```

Comment:
if we have data queued and we received a terminal event, we should deliver the queued elements when we get more demand.

```diff
diff --git a/servicetalk-concurrent-api/src/test/java/io/servicetalk/concurrent/api/PublisherBufferTest.java b/servicetalk-concurrent-api/src/test/java/io/servicetalk/concurrent/api/PublisherBufferTest.java
index acf6b8a4c..b951a05ff 100644
--- a/servicetalk-concurrent-api/src/test/java/io/servicetalk/concurrent/api/PublisherBufferTest.java
+++ b/servicetalk-concurrent-api/src/test/java/io/servicetalk/concurrent/api/PublisherBufferTest.java
@@ -25,6 +25,7 @@ import org.hamcrest.Matcher;
 import org.junit.jupiter.api.Test;
 
 import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.Future;
 import java.util.concurrent.LinkedBlockingDeque;
 import java.util.concurrent.atomic.AtomicInteger;
 import java.util.concurrent.atomic.AtomicReference;
@@ -48,17 +49,63 @@ class PublisherBufferTest {
 
     private static final int EMPTY_ACCUMULATOR_VAL = -1;
     static final int BUFFER_SIZE_HINT = 8;
-    private final TestPublisher<Integer> original = new TestPublisher<>();
-    private final TestPublisher<Accumulator<Integer, Integer>> boundaries = new TestPublisher<>();
+    private final TestSubscription tSubscription = new TestSubscription();
+    private final TestPublisher<Integer> original = new TestPublisher.Builder<Integer>().disableAutoOnSubscribe()
+            .build(sub -> {
+                sub.onSubscribe(tSubscription);
+                return sub;
+            });
+    private final TestSubscription bSubscription = new TestSubscription();
+    private final TestPublisher<Accumulator<Integer, Integer>> boundaries =
+            new TestPublisher.Builder<Accumulator<Integer, Integer>>().disableAutoOnSubscribe().build(sub -> {
+                sub.onSubscribe(bSubscription);
+                return sub;
+            });
     private final TestPublisherSubscriber<Integer> bufferSubscriber = new TestPublisherSubscriber<>();
 
-    PublisherBufferTest() {
+    PublisherBufferTest() throws Exception {
         toSource(original.buffer(new TestBufferStrategy(boundaries, BUFFER_SIZE_HINT))).subscribe(bufferSubscriber);
         bufferSubscriber.awaitSubscription().request(1); // get first boundary
+        bSubscription.awaitRequestN(1);
         boundaries.onNext(new SumAccumulator());
         assertThat(bufferSubscriber.pollOnNext(10, MILLISECONDS), is(nullValue()));
     }
 
+    @Test
+    void terminalBeforeRequestedDelivers() throws Exception {
+        Executor executor = Executors.newCachedThreadExecutor();
+        try {
+            Future<Void> f = executor.submit(() -> {
+                try {
+                    tSubscription.awaitRequestN(3);
+                } catch (Exception e) {
+                    throw new AssertionError(e);
+                }
+
+                original.onNext(1);
+
+                try {
+                    bSubscription.awaitRequestN(2);
+                } catch (Exception e) {
+                    throw new AssertionError(e);
+                }
+                boundaries.onNext(new SumAccumulator());
+
+                original.onNext(2);
+                original.onComplete();
+            }).toFuture();
+
+            assertThat(bufferSubscriber.takeOnNext(), is(1));
+
+            f.get();
+            bufferSubscriber.awaitSubscription().request(1);
+            assertThat(bufferSubscriber.takeOnNext(), is(2));
+            bufferSubscriber.awaitOnComplete();
+        } finally {
+            executor.closeAsync().toFuture().get();
+        }
+    }
+
     @Test
     void invalidBufferSizeHint() {
         TestPublisherSubscriber<Integer> bufferSubscriber = new TestPublisherSubscriber<>();

```